File: leetcode/js/add-binary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function (a, b) {
  a = [...a].reverse().join("");
  b = [...b].reverse().join("");

  let result = "";

  let aIdx = 0;
  let bIdx = 0;
  let carry = 0;

  while (aIdx < a.length || bIdx < b.length || carry !== 0) {
    let sum = Number(a[aIdx] || 0) + Number(b[bIdx] || 0) + carry;

    if (sum > 1) {
      sum %= 2;
      carry = 1;
    } else {
      carry = 0;
    }

    result = `${sum}${result}`;

    aIdx++;
    bIdx++;
  }

  return result;
};

var addBinary = function (a, b) {
  const sum = BigInt(`0b${a}`) + BigInt(`0b${b}`);

  return sum.toString(2);
};

File: leetcode/rust/add-binary.rs
impl Solution1 {
    pub fn add_binary(a: String, b: String) -> String {
        let a = a.chars().rev().collect::<Vec<_>>();
        let b = b.chars().rev().collect::<Vec<_>>();

        let mut result = String::new();

        let mut a_idx = 0;
        let mut b_idx = 0;
        let mut carry = 0;

        while a_idx < a.len() || b_idx < b.len() || carry != 0 {
            let a_val = a.get(a_idx).unwrap_or(&'0');
            let b_val = b.get(b_idx).unwrap_or(&'0');

            let mut sum = a_val.to_digit(10).unwrap() + b_val.to_digit(10).unwrap() + carry;

            if sum > 1 {
                sum = sum % 2;
                carry = 1;
            } else {
                carry = 0;
            }

            result = format!("{sum}{result}");

            a_idx += 1;
            b_idx += 1;
        }

        result
    }
}

impl Solution2 {
    pub fn add_binary(a: String, b: String) -> String {
        let a_num = u128::from_str_radix(&a, 2).unwrap();
        let b_num = u128::from_str_radix(&b, 2).unwrap();

        let sum = a_num + b_num;

        format!("{:b}", sum)
    }
}
